File: frontend/common/lib/models/teacher_model.dart
import 'package:json_annotation/json_annotation.dart';

import 'user_model.dart';

part 'teacher_model.g.dart';

@JsonSerializable(
  nullable: false,
  fieldRename: FieldRename.none,
)
class TeacherModel {
  @JsonKey(name: '_id')
  String id;
  @JsonKey(ignore: true)
  UserModel userAccount = new UserModel();
  // personal details
  String gender;
  String full_name;
  String date_of_birth;
  String blood_group;
  String phone;
  String qualification;
  String subjects_handling;
  String incharge_class;
// address:
  String street_address;
  String city_name;
  String country;
  String pin_code;

// school details:
  String joining_date;
  String leaving_date;
  String current_position;
  String employee_code;
  String working_hours;

  TeacherModel({
    this.id,
    this.gender,
    this.full_name,
    this.date_of_birth,
    this.blood_group,
    this.phone,
    this.qualification,
    this.street_address,
    this.city_name,
    this.country,
    this.pin_code,
    this.joining_date,
    this.leaving_date,
    this.current_position,
    this.employee_code,
    this.working_hours,
    this.subjects_handling,
    this.incharge_class,
  });

  factory TeacherModel.fromJson(Map<String, dynamic> json) {
    TeacherModel teacherModel = _$TeacherModelFromJson(json);
    teacherModel.userAccount = UserModel.fromJson(json["user"]);
    return teacherModel;
  }
  Map<String, dynamic> toJson() {
    Map<String, dynamic> map = _$TeacherModelToJson(this);
    map["user"] = userAccount.id;
    return map;
  }
}

File: backend/src/models/users/teacher.model.js
const mongoose = require('mongoose');

const teacherSchema = new mongoose.Schema({
    createdAt: {
        type: Date,
        default: Date.now()
    },
    user: {
        type: mongoose.Schema.ObjectId,
        ref: 'User',
        required: [true, 'teacher must belong to a User account!']
      },
      name: {
        type: String,
        required: [true, 'A personal must have a name'],
    },
    gender: {
        type: String,
        required: [true, 'A personal must have a gender'],
        enum: {
            values: ['male', 'female'],
            message: 'Gender is either: male or female'
        }
    },
    date_of_birth: {
        type: Date,
        // required: [true, 'A personal must have a date of birth'],
    },
    qualification: {
        type: String,
    },
    phone: {
        type: String,
    },
    street_address: {
        type: String,
    },
    city_name: {
        type: String,
    },
    country: {
        type: String,
    },
    pin_code: {
        type: String,
    },
    joining_date: {
        type: String,
    },
    leaving_date: {
        type: String,
    },
    current_position: {
        type: String,
    },
    employee_code: {
        type: String,
    },
},
    {
        toJSON: { virtuals: true },
        toObject: { virtuals: true }
    });

teacherSchema.pre(/^find/, function (next) {
    this.populate({
        path: 'user',
        select: 'name email photo'
    });
    next();
});

const userModel = require('./user.model');
teacherSchema.post('findOneAndDelete',async function(doc){
    if (doc.user._id) {
        const user = this
        // await userModel.findOneAndDelete({_id:doc.user._id})
        await userModel.deleteOne({_id:doc.user._id})

    }
})
// https://stackoverflow.com/questions/11904159/automatically-remove-referencing-objects-on-deletion-in-mongodb
// teacherSchema.pre('findOneAndDelete',async function(next){
//     const doc = this
//     if (doc.user._id) {
      
//         // await userModel.findOneAndDelete({_id:doc.user._id})
//         // await userModel.deleteOne({_id:doc.user._id})
//         person.model('User').deleteOne({ _id: doc.user._id }, next);

//     }
// })

const Teacher = mongoose.model('teachers', teacherSchema);

module.exports = Teacher;
